File: backend/app/schemas/stats.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class GameStatCreateRequest(BaseModel):
    session_id: str = Field(..., min_length=8, max_length=128)
    partei: str = Field(..., min_length=2, max_length=16)
    complexity: int = Field(..., ge=1, le=4)
    gewonnen: bool
    wahlprognose: float = Field(..., ge=0, le=100)
    monate_gespielt: int = Field(..., ge=1, le=48)
    gesetze_beschlossen: int | None = Field(None, ge=0)
    gesetze_gescheitert: int | None = Field(None, ge=0)
    koalitionsbruch: bool = False
    saldo_final: float | None = None
    gini_final: float | None = Field(None, ge=0, le=100)
    arbeitslosigkeit_final: float | None = Field(None, ge=0, le=100)
    medienklima_final: int | None = Field(None, ge=0, le=100)
    skandale_gesamt: int | None = Field(None, ge=0)
    pk_verbraucht: int | None = Field(None, ge=0)
    top_politikfeld: str | None = Field(None, max_length=64)
    bewertung_gesamt: str | None = Field(None, max_length=1)
    titel: str | None = Field(None, max_length=128)
    opt_in_community: bool = False


class GameStatCreateResponse(BaseModel):
    id: str
    recorded: bool = True


class NachParteiRow(BaseModel):
    partei: str
    anzahl: int
    gewinnrate: float
    wahlprognose_avg: float


class PolitikfeldCount(BaseModel):
    feld: str
    anzahl: int


class TitelCount(BaseModel):
    titel: str
    anzahl: int


class CommunityStatsResponse(BaseModel):
    gesamt: int
    gewinnrate: float
    wahlprognose_avg: float
    nach_partei: list[NachParteiRow]
    top_politikfelder: list[PolitikfeldCount]
    titel_verteilung: list[TitelCount]


class MeStatItem(BaseModel):
    id: str
    partei: str
    complexity: int
    gewonnen: bool
    wahlprognose: float
    monate_gespielt: int
    bewertung_gesamt: str | None
    titel: str | None
    created_at: datetime


class HighscoreItem(BaseModel):
    titel: str | None
    partei: str
    wahlprognose: float
    gesetze_beschlossen: int | None
    saldo_final: float | None
    complexity: int
    created_at: datetime


class HighscoresResponse(BaseModel):
    items: list[HighscoreItem]

File: frontend/src/services/stats.ts
import { apiFetch } from './api';

const STATS_SESSION_KEY = 'br_stats_session_v1';

/** Fallback wenn localStorage nicht verfügbar — stabil pro App-Session */
let fallbackSessionId: string | null = null;

/** Stabile anonyme Session für Community-Stats (ohne Login) */
export function getOrCreateStatsSessionId(): string {
  try {
    let id = localStorage.getItem(STATS_SESSION_KEY);
    if (!id) {
      id = crypto.randomUUID();
      localStorage.setItem(STATS_SESSION_KEY, id);
    }
    return id;
  } catch {
    // localStorage nicht verfügbar — ID pro App-Session stabil halten
    if (!fallbackSessionId) {
      fallbackSessionId = crypto.randomUUID();
    }
    return fallbackSessionId;
  }
}

export interface GameStatPayload {
  session_id: string;
  partei: string;
  complexity: number;
  gewonnen: boolean;
  wahlprognose: number;
  monate_gespielt: number;
  gesetze_beschlossen?: number;
  gesetze_gescheitert?: number;
  koalitionsbruch: boolean;
  saldo_final?: number;
  gini_final?: number;
  arbeitslosigkeit_final?: number;
  medienklima_final?: number;
  skandale_gesamt?: number;
  pk_verbraucht?: number;
  top_politikfeld?: string | null;
  bewertung_gesamt?: string;
  /** SMA-499: Gesamtpunkte 0–100 (Bilanz/Agenda/Urteil + Wahlbonus) */
  spielziel_gesamtpunkte?: number;
  titel?: string;
  opt_in_community: boolean;
}

export async function postGameStats(
  payload: GameStatPayload,
  accessToken?: string | null,
): Promise<{ id: string }> {
  return apiFetch<{ id: string }>('/stats', {
    method: 'POST',
    body: payload,
    ...(accessToken ? { token: accessToken } : {}),
  });
}

export interface CommunityStats {
  gesamt: number;
  gewinnrate: number;
  wahlprognose_avg: number;
  nach_partei: Array<{
    partei: string;
    anzahl: number;
    gewinnrate: number;
    wahlprognose_avg: number;
  }>;
  top_politikfelder: Array<{ feld: string; anzahl: number }>;
  titel_verteilung: Array<{ titel: string; anzahl: number }>;
}

export async function fetchCommunityStats(): Promise<CommunityStats> {
  return apiFetch<CommunityStats>('/stats/community');
}

export interface HighscoreItem {
  titel: string | null;
  partei: string;
  wahlprognose: number;
  gesetze_beschlossen: number | null;
  saldo_final: number | null;
  complexity: number;
  created_at: string;
}

export async function fetchHighscores(partei?: string, complexity?: number, limit = 20) {
  const q = new URLSearchParams();
  if (partei) q.set('partei', partei);
  if (complexity != null) q.set('complexity', String(complexity));
  q.set('limit', String(limit));
  const qs = q.toString();
  return apiFetch<{ items: HighscoreItem[] }>(`/stats/highscores${qs ? `?${qs}` : ''}`);
}
